parallel-install/pkg/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/kyma-incubator/hydroform/parallel-install/pkg/logger"
+	"k8s.io/client-go/kubernetes"
+)
+
+type component string
+type executionTime int
+type jobName string
+
+type jobStatus struct {
+	job    jobName
+	status bool
+	err    error
+}
+
+type job interface {
+	execute(*config.Config, kubernetes.Interface, context.Context) error
+	when() (component, executionTime)
+	identify() jobName
+}
+
+const (
+	Pre executionTime = iota
+	Post
+)
+
+var duration time.Duration
+var durationMu sync.Mutex
+
+var preJobMap = make(map[component][]job)
+var postJobMap = make(map[component][]job)
+var finishedJobs = []jobStatus{}
+
+var kubeClient kubernetes.Interface
+var cfg *config.Config
+
+var log logger.Interface
+
+// Register job
+func register(j job) int {
+	component, executionTime := j.when()
+	if executionTime == Pre {
+		preJobMap[component] = append(preJobMap[component], j)
+	} else if executionTime == Post {
+		postJobMap[component] = append(postJobMap[component], j)
+	}
+	return 0
+}
+
+// Sets Installation Config and KubeClient at package level
+func RegisterJobManager(config *config.Config, kc kubernetes.Interface) {
+	cfg = config
+	kubeClient = kc
+}
+
+// Sets Logger at package level
+func SetLogger(logClient logger.Interface) {
+	log = logClient
+}
+
+// Function should be called before component is being deployed/upgraded
+// If the Context is cancelled, the worker quits immediately, skipping the remaining components
+func ExecutePre(ctx context.Context, c string) {
+	execute(ctx, c, preJobMap)
+}
+
+// Function should be called after compoent is being deployed/upgraded
+// If the Context is cancelled, the worker quits immediately, skipping the remaining components
+func ExecutePost(ctx context.Context, c string) {
+	execute(ctx, c, postJobMap)
+}
+
+// Used by ExecutePre() && ExecutePost()
+// Used to start workers and grab jobs belonging to the respective component
+func execute(ctx context.Context, c string, executionMap map[component][]job) {
+	var wg sync.WaitGroup
+
+	start := time.Now()
+
+	jobs := executionMap[component(c)]
+	statusChan := make(chan jobStatus, len(jobs))
+	wg.Add(len(jobs))
+
+	if len(jobs) > 0 {
+		for _, job := range jobs {
+			go worker(ctx, statusChan, &wg, job)
+		}
+	}
+	go func() {
+		wg.Wait()
+		close(statusChan)
+	}()
+
+	emptyJob := jobStatus{}
+	for status := range statusChan {
+		log.Infof("Job Status: %v", status)
+		if status != emptyJob {
+			finishedJobs = append(finishedJobs, status)
+			if status.status == true {
+				log.Infof("Following job executed: %v", status.job)
+			} else if status.status == false {
+				log.Infof("Following job failed while execution: `%v` with error: %s", status.job, status.err)
+			}
+		}
+	}
+
+	t := time.Now()
+	addDuration(t.Sub(start))
+}
+
+func worker(ctx context.Context, statusChan chan<- jobStatus, wg *sync.WaitGroup, j job) {
+	defer wg.Done()
+	if err := j.execute(cfg, kubeClient, ctx); err != nil {
+		j := jobStatus{j.identify(), false, err}
```

Comment:
call the variable `status` as j is already defined as job and you are creating a new variable in if block. Could be a bit confusing.